### DISCUSSION TITLE
ProMicroの認識について

### DISCUSSION THREAD

**Discussion by akamini** [Yowkees/keyball#391] [Q&A]
# 前提条件
 左手側にtype b端子、右手側にtype c端子のProMicroをそれぞれハンダ付けしています。

## 現状
REMAPを使用し、それぞれのProMicroで動作確認(Test Matrix Mode)を行いました。
右手側(type c)はキーの位置が正常に認識されていますが、左手側(type b)が反転して認識されます。
右手側にトラックボールを配置しているのですが、左手側にトラックボールがあると認識されます。
また、トラックボールの動作も反転しています。

## 質問したい内容
- 上記の問題はREMAPやファームウェアの設定次第で修正できるものでしょうか？
- REMAPを使用してキーマップを右手側、左手側と設定したい際にどちらのProMicroに接続しても同一のキーボードと認識されていますが、こちらは問題ないのでしょうか？
- ProMicroのハンダ不良によってこれらの問題が起きる可能性はあるのでしょうか？

右手側・左手側に同じキーマップを配置しています。
正常に動作している右手側
![1697466807552](https://github.com/Yowkees/keyball/assets/68289484/ec4ef0db-45fb-408a-a554-4719be04d164)
キーボードのスイッチの配置とREMAPの配置が一致しています。
トラックボールは右側に回すと右にカーソルが進みます。

正常に動作していない左手側
![1697643492896@2x](https://github.com/Yowkees/keyball/assets/68289484/10c2c946-7711-47f3-8bf1-d7f4f617a368)
キーボードのスイッチの配置とREMAPの配置が反転しています。
トラックボールは右側に回すと左にカーソルが進みます。

画像はTest Matrix Modeの画像で「Y・H・N」を押した時に左手側の「T・G・B」の箇所が点灯し、「Esc・Tab・Shift」を押した時に右手側の「Del・＆・１」の箇所が点灯します。

**Comment by Yowkees** [comment]
ご連絡ありがとうございます。

私も初めて聞いた症状です。ちゃんと左右共キーが反応するので左右の通信には問題なさそうですね。その上で配置が反転するというの初めてです。

年のため正規ファームウェアを再度書き込み直してみてください。
また、左右のProMicroを入れ替えるとどのような挙動を示すか試してみてください。

以上、お手数ですがご確認お願いいたします。

**Reply by akamini** [reply]
書き直し及び入れ替えを行ってみましたが、挙動としては同じ挙動を示すことが確認できました。

## 試した内容
- Keyball44_via_v1.2.0を左右それぞれ片方ずつ再度書き込み
  - 挙動に変化なし
- ProMicroを入れ替え、Keyball44_via_v1.2.0を左右それぞれ片方ずつ書き込み
  - 挙動に変化なし

## 確認した点
- コンスルーの向き
- ジャンパのハンダ付け
  - これらに問題がある際にはキーを認識しないと思っていますが、そのような認識で間違いないでしょうか？  

テスターがない環境のため、複雑な確認はできませんがProMicroを入れ替えても同じ現象が起きているのでProMicroやファームウェア川の問題ではないと考えています。

**Comment by Yowkees** [comment]
ご確認ありがとうございます。

可能性として、ProMicroの差し込み位置が違うということはありませんでしょうか？
ミドル基板の黒枠で示した個所に差し込まれているかご確認ください。

解決しないようでしたら、何かわかるかも知れませんので半田付け面のお写真を見せていただければ幸いです。

以上、お手数ですがご対応お願いいたします。

**Reply by akamini** [reply]
両方とも黒枠で囲われた部分に差し込まれていることを確認いたしました。

半田付け面の画像になります。
全体を撮影しております。
こちらが左手側です。
![S__7594000_0](https://github.com/Yowkees/keyball/assets/68289484/9c0a0289-558a-48c8-be1d-71bdce5ce5bc)
こちらが右手側です。
![S__7593998_0](https://github.com/Yowkees/keyball/assets/68289484/047780ee-b7c7-4b01-8f76-5e0b0854e945)

試しにトラックボール基盤を外して左手側、右手側にそれぞれ接続したところ、どちらに接続しても下記のようにトラックボールなし判定されましたが、TestMatrixModeの挙動は同じでした。
![1697807550912@2x](https://github.com/Yowkees/keyball/assets/68289484/88fb782b-89ce-4893-b384-7cf61c3849a8)

---

#### 私のPC環境
MacBook Pro 13-inch 2019

メモリ8GB

**Comment by Yowkees** [comment]
ProMicroを左右入れ替えた場合でも、
・右手側にUSB接続時は正常
・左手側にUSB接続時は左右反転
という症状で合っていますでしょうか？

**Reply by akamini** [reply]
その認識で正しいです。

WindowsPCで再書き込みして試してみましたが、特に変化なしでした。
